GeomIO/SharedIO.F90

```diff
@@ -13,9 +14,128 @@ module mapl3g_SharedIO
    public get_mapl_geom
    public create_time_variable
    public bundle_to_metadata
+   public esmf_to_pfio_type
+   public create_local_start
+   public create_global_count
+   public create_global_start
+   public create_file_shape
 
    contains
 
+   function create_file_shape(grid, field_shape, rc) result(file_shape)
+      integer, allocatable :: file_shape(:)
+      type(ESMF_Grid), intent(in) :: grid
+      integer, intent(in) :: field_shape(:)
+      integer, intent(out), optional :: rc
+
+      integer :: status, sz, ungr, tile_count
+      call ESMF_GridGet(grid, tileCount=tile_count, _RC)
+      sz = size(field_shape)
+      ungr = sz - 2
+      if (tile_count == 6) then
+         allocate(file_shape(sz+1))
+         file_shape(1:3) = [field_shape(1), field_shape(2), 1]
+         file_shape(4:4+ungr-1) = [field_shape(2+ungr:sz)]
+      else if (tile_count == 1) then
+         file_shape = field_shape
+      else 
+         _FAIL("unsupported grid")
+      end if
+
+      _RETURN(_SUCCESS)
+   end function create_file_shape
+
+   function create_global_start(grid, field_shape, time_index, rc) result(global_start)
+      integer, allocatable :: global_start(:)
+      type(ESMF_Grid), intent(in) :: grid
+      integer, intent(in) :: field_shape(:)
+      integer, optional,  intent(in) :: time_index
+      integer, intent(out), optional :: rc
+
+      integer :: status, sz, tile_count, tm
+      call ESMF_GridGet(grid, tileCount=tile_count, _RC)
+      sz = size(field_shape)
+
+      tm = 0
+      if (present(time_index)) tm=1
+      if (tile_count == 6) then
+         allocate(global_start(sz+1+tm))
+         global_start(1:sz+1) = 1
+         if (present(time_index)) global_start(sz+2) = time_index
+      else if (tile_count == 1) then
+         allocate(global_start(sz+tm))
+         global_start(1:sz) = 1
+         if (present(time_index)) global_start(sz+1) = time_index
+      else 
+         _FAIL("unsupported grid")
+      end if
```

Comment:
As you might expect, this trips my `else` klaxon ...

Probably none of the options are great, but the duplicated checks on the presence of `time_index` likely pushes toward some sort of refactor.    Could possibly due this as a pair-programming exercise to reduce latency in the likely event multiple iterations are needed.